lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb

```diff
@@ -0,0 +1,202 @@
+require "topological_inventory/ansible_tower/collector"
+require "topological_inventory/providers/common/operations/sources_api_client"
+require "topological_inventory/ansible_tower/receptor/async_receiver"
+
+module TopologicalInventory
+  module AnsibleTower
+    module TargetedRefresh
+      class ServiceInstance < TopologicalInventory::AnsibleTower::Collector
+        REFS_PER_REQUEST_LIMIT = 20
+
+        def initialize(payload = {})
+          self.params    = payload['params']
+          self.source_id = payload['source_id']
+
+          # TODO: add metrics exporter
+          super(payload['source_uid'], nil)
+        end
+
+        # Entrypoint for 'ServiceInstance.refresh' operation
+        def refresh
+          return if params_missing?
+
+          # Connection settings
+          set_connection_data!
+          logger.info("ServiceInstance#refresh - Connection set successfully")
+
+          # Get all tasks
+          tasks, refresh_state_uuid = {}, SecureRandom.uuid
+          params.to_a.each do |task|
+            if task['task_id'].blank? || task['source_ref'].blank?
+              logger.warn("ServiceInstance#refresh - missing data for task: #{task}")
+              next
+            end
+            tasks[task['task_id']] = task['source_ref']
+
+            if tasks.length == REFS_PER_REQUEST_LIMIT
+              refresh_part(tasks, refresh_state_uuid, SecureRandom.uuid)
+              tasks = {}
+            end
+          end
+
+          refresh_part(tasks, refresh_state_uuid, SecureRandom.uuid) unless tasks.empty?
+        rescue => err
+          logger.error("ServiceInstance#refresh - Error: #{err.message}\n#{err.backtrace.join("\n")}")
```

Comment:
Can we get some metrics around this? Would be really nice to alert when `ServiceInstance#refresh` is failing a lot.

Reply:
I guess I added some TODO, I think this PR is big enough, so I'll rather add metrics in next PR